### DISCUSSION TITLE
6주차: 9장 데이터 조작화

### DISCUSSION THREAD

**Discussion by choisohyun** [frontend-book-study/refactoring-2#13] [리팩터링 2판]
인상깊었던 책의 내용과 인상깊었던 이유
현업 등 코드에 반영할 수 있는 방법과 생각
관련 아티클이나 이해를 돕는 자료를 첨부
새롭게 알게된 점이나 이해가 어려웠던 점

**Comment by choisohyun** [comment]
변수 쪼개기에서 '입력 매개변수의 값을 수정할 때' if문 코드의 경우 let을 사용하지 않고 바로 반환하면 더 보기 편할 것 같다.

다른 부분은 일반적으로 진행하곤 하는 리팩터링 방법인 것 같다!

**Comment by LastCow9000** [comment]
9장을 요약하자면 사이드 이펙트를 줄이기 위해 값을 불변으로 다루는 것의 중요성을 알려주는 내용인 것 같습니다. 예외적으로 9.5장 값을 참조로 바꾸기는 값의 변경이 전파되기를 원하는 곳에 적용해야 하므로 주의해서 사용해야 할 것 같습니다.
또한 개발하다보면 매직 리터럴로 바꾸는 것을 까먹고 하드코딩 할 때가 많은데 다시 리마인드 하게 되었습니다.

**Comment by RAINBOWJYP** [comment]
> 하나의 값이 여러 목적으로 사용된다면 혼란과 버그를 낳는다.

React.useState를 쓸 때 한번만 대입해야겠다는 생각을 하게되었다.

책을 읽으면서 리팩터링 방법 초입부에 그려진 그림을 눈여겨보게되었다.
이해하기 쉽게 그려져있다는걸 스터디 중반부에 알게되었다. 
참조를 값으로 바꾸기, 값을 참조로 바꾸기 부분을 그림으로 보니 훨씬 이해가 쉬웠다.

**Comment by hyungwookchoi** [comment]
> '작은 단계'라 함은 각 단계에서 잘못될 일이 적어졌고, 그래서 할 일도 줄었다는 뜻이다. p.335

단계를 나누는 것이 구조를 복잡하게 만드는 것처럼 보일 수 있지만, 실제로는 각 단계에서 잘못 될 일이 적어졌고 실수 할 일이 적어졌기 때문에 더 쉬워지는 것이라는 것을 느꼈습니다.

**Comment by hamo-o** [comment]
> 9.1 변수 쪼개기 - 대입이 두 번 이상 이뤄진다면 여러 가지 역할을 수행한다는 신호다. 역할이 둘 이상인 변수가 있다면 쪼개야 한다. 예외는 없다.

"예외는 없다" 라는 단호한 표현이 인상깊었습니다. 개발할 때 역할이 여러개인 변수가 없는지 잘 확인해봐야겠습니다.

> 9.4 참조를 값으로 바꾸기, 9.5 값을 참조로 바꾸기 - 논리적으로 같은 데이터를 물리적으로 복제해 사용할 때 가장 크게 문제되는 상황은 그 데이터를 갱신해야 할 때다.

불변성을 위해 참조를 값으로 바꾸는 작업은 많이 했는데, 그 반대는 잘 활용하지 못했던 것 같습니다. 갱신되는 값들이 여러 곳에 퍼져있을 때 활용해봐야겠습니다.

**Comment by blan19** [comment]
대부분의 에러는 메모리 참조에서 비롯되는 것 처럼

`값을 참조로 바꾸기, 참조를 값으로 바꾸기` 이 두가지를 적절하게 다루는 것은 매우 중요한 것 같습니다.

복잡한 어플리케이션의 경우 참조 객체를 잘못 다루면 금방 복잡해지기 때문에 무조건 불변성을 지켜 무분별하게 코드를 작성하는 경우가 간혹 있었는데, 그 역을 잘 활용해 쓸모 없는 코드를 제거하는 노력도 해봐야겠습니다!

**Comment by suojae** [comment]
1. "데이터 구조의 필드 이름은 이해의 핵심이다"
2. 값을 선언함에 있어 동기화를 할지 불변성을 유지할지 고려해야겠다는 생각이 들었습니다.
3. 한 번만 쓰는 변수는 선언하지 말고 함수로 chaining해야겠다는 생각이 들었습니다.
